plugins/outputs/graylog/graylog.go

```diff
@@ -21,45 +24,80 @@ const (
 	defaultConnection      = "wan"
 	defaultMaxChunkSizeWan = 1420
 	defaultMaxChunkSizeLan = 8154
+	defaultScheme          = "udp"
+	defaultTimeout         = 5 * time.Second
 )
 
-type GelfConfig struct {
+type gelfConfig struct {
 	GraylogEndpoint string
 	Connection      string
 	MaxChunkSizeWan int
 	MaxChunkSizeLan int
 }
 
-type Gelf struct {
-	GelfConfig
+type gelf interface {
+	io.WriteCloser
 }
 
-func NewGelfWriter(config GelfConfig) *Gelf {
-	if config.GraylogEndpoint == "" {
-		config.GraylogEndpoint = defaultGraylogEndpoint
+type gelfCommon struct {
+	gelfConfig
+	dialer *net.Dialer
+	conn   net.Conn
+}
+
+type gelfUDP struct {
+	gelfCommon
+}
+
+type gelfTCP struct {
+	gelfCommon
+}
+
+func newGelfWriter(cfg gelfConfig, dialer *net.Dialer) gelf {
+	if cfg.GraylogEndpoint == "" {
+		cfg.GraylogEndpoint = defaultGraylogEndpoint
 	}
 
-	if config.Connection == "" {
-		config.Connection = defaultConnection
+	if cfg.Connection == "" {
+		cfg.Connection = defaultConnection
 	}
 
-	if config.MaxChunkSizeWan == 0 {
-		config.MaxChunkSizeWan = defaultMaxChunkSizeWan
+	if cfg.MaxChunkSizeWan == 0 {
+		cfg.MaxChunkSizeWan = defaultMaxChunkSizeWan
 	}
 
-	if config.MaxChunkSizeLan == 0 {
-		config.MaxChunkSizeLan = defaultMaxChunkSizeLan
+	if cfg.MaxChunkSizeLan == 0 {
+		cfg.MaxChunkSizeLan = defaultMaxChunkSizeLan
 	}
 
-	g := &Gelf{GelfConfig: config}
+	scheme := defaultScheme
+	parts := strings.SplitN(cfg.GraylogEndpoint, "://", 2)
+	if len(parts) == 2 {
+		scheme = strings.ToLower(parts[0])
+		cfg.GraylogEndpoint = parts[1]
+	}
+	common := gelfCommon{
+		gelfConfig: cfg,
+		dialer:     dialer,
+	}
+
+	var g gelf
+	switch scheme {
+	case "tcp":
+		g = &gelfTCP{gelfCommon: common}
+	case "udp":
+		fallthrough
```

Comment:
why not just have the default if this is fallthrough?

Reply:
Removed.